### ISSUE TITLE
IOControl_SIOCATMARK_Success test not working on macOS/Windows

### ISSUE THREAD

**Issue by stephentoub** [dotnet/corefx#25639]
The new IOControl_SIOCATMARK_Success is failing on macOS.  Needs investigation to understand whether it's a test issue (e.g. a subtle behavioral difference between unix implementations) or something deeper.

**Comment by stephentoub** [comment]
It also failed sporadically on Windows.

**Comment by rmkerr** [comment]
So I've been looking at this test and reading up on how we work with out-of-band data. I have some ideas, but can't find any documentation of the behavior difference I'm seeing.

The failing test follows this process:
1. Set up and connect a client and a server socket, both using TCP.
2. Send four bytes of regular data (integer value 42).
3. Send four bytes of out-of-band data (integer value 43).
4. Check that the TCP stream is not yet at the out-of-band mark.
5. Read from the TCP stream, check that the read value is 42.
6. Read the out-of-band data, check that the read value is 43.
7. Check that the TCP stream _is_ at the out-of-band mark.

When run on OSX the test consistently fails the check at step 7, indicating that our current location within the TCP stream is not the location of the out-of-band mark. To test where the mark might actually be located, I added logging to the test that would check the status of the mark between each of the steps above:
```
After send: 0
After first receive: 1
After second receive: 0
```
The logging indicated the out-of-band mark is being hit between steps 5 and 6. This means that after the read from the out-of-band data (which is separate from the TCP stream), the check indicates that the position of the TCP stream no longer matches the position of the out-of-band mark. Since the TCP stream position has not changed, it seems that the out-of-band mark must have been modified. That leads me to the following theory: **OSX clears the out-of-band mark once out-of-band data has been read.** 

If that is the case I think this is the kind of behavior difference we would just have to document and work around. It wouldn't be hard to fix the test -- we'd just have to switch steps 6 and 7.

Thoughts? I'm not sure who is the expert on this kind of low level sockets issue.

**Comment by rmkerr** [comment]
Hey @davidsh, do you mind taking a look at this or pointing me to someone who could? I'm basing my analysis off of some very limited documentation, and it'd be nice to have an expert take a look before I write it off as just a behavior difference.

**Comment by rmkerr** [comment]
I've looked into this issue in a lot more detail and have talked to some of the area experts. In summary it looks like the Windows and the Unix implementations of SIOCATMARK are _completely_ different. I don't know if we should document the difference, or if we should just not support these cross platform.

Once a socket has been initialized but before data has been sent, the Unix implementation has a value of 0. The windows implementation has a value of 1. Both keep the same value when regular (non-oob) data is sent.

Once out-of-band data has been sent, the Windows implementation of SIOCATMARK changes to the value 0, regardless of the current read position in the TCP stream. The value will remain at 0 until the out-of-band data is read, and it then switches back to 1. In contrast the value of SIOCATMARK on Unix remains 0 until the TCP read stream pointer reaches the point in the stream where the out of band data would have been had it not been separated from the stream. It then takes on a value of 1, which it keeps until the position of the TCP read stream pointer changes.

Here's a table to make the differences more obvious:

  | SIOCATMARK(Win) | SIOCATMARK(Unix)
-- | -- | --
Before Data   Received: | 1 | 0
After Non-OOB Data   Received | 1 | 0
After OOB Data   Received (Read pointer not at OOB Mark) | 0 | 0
After OOB Data   Received (Read pointer at OOB Mark) | 0 | 1
OOB Data Read   (Read pointer at OOB Mark) | 1 | (1/0)
OOB Data Read   (Read pointer not at OOB Mark) | 1 | 0

Note: When the OOB data has been read but the read pointer is still at the OOB mark, behavior varies on Linux and OSX.

In conclusion: the flag has a completely different meaning between operating systems. In Windows it appears that it is used to test for the presence of OOB data in the stream, regardless of the position. On Unix it is used to test if the read pointer of the TCP stream matches point in the where the out of band data would have been had it not been separated from the regular stream.

**Comment by davidsh** [comment]
Good analysis!  In terms of next steps, I think we should simply document this (perhaps using your details above).  

What sort of applications/services would be broken by just leaving things as-is and documenting? I don't see how inventing an aligned behavior helps real apps.  So, I'm inclined to just document the platform difference.

**Comment by stephentoub** [comment]
That's very helpful.  Thanks, @rmkerr.

> I think we should simply document this

+1, basically you get whatever functionality the OS provides here.  Ideally we'd update/split the test so that we can enable it and run it in CI.

**Comment by rmkerr** [comment]
That sounds good to me. I'll send out a PR with split up tests for OS specific behavior. 

@davidsh As far as I know this isn't a widely used feature, and when it is used it is to interface with legacy systems that may have their own behavior for tcp urgent. I think that the use case is advanced enough that developers are going to take the platform specific behavior into consideration. It definitely needs documentation though, and I can look into adding that too.

**Comment by rmkerr** [comment]
Closing since this test was reworked in #27583.
